File: src/ocaml/domain/notification.ml
(** [Notification] provides types for notification that send to client. *)

(** Level of notification. *)
module Level = struct
  type t =
    | Info
    | Warning
    | Error

  let to_int = function Info -> 0 | Warning -> 1 | Error -> 2
  let of_int = function 0 -> Some Info | 1 -> Some Warning | 2 -> Some Error | _ -> None
end

(** [body] describes body of the notification. *)
type body =
  | Message of string
  | Progress of {process : string; current : float; targeted : float}

(** Identifier of the notification. Each notifications has global unique identifier.  *)
type id = Uuidm.t

type t =
  { id : id
  ; level : Level.t
  ; body : body }

let make ~id ~level ~body = {id; level; body}

module type Factory = sig
  val create : level:Level.t -> body:body -> t
  (** [create ~level ~body] create a instance of [t]. *)
end

File: src/ts/domain/notification.ts
// type of notification
export enum NotificationType {
  OneShot = "oneshot",
  Progress = "progress",
}

export enum Level {
  Info = "info",
  Warning = "warning",
  Error = "error",
}

export interface OneShotBody {
  kind: "oneshot";
  message: string;
}

export interface ProgressBody {
  kind: "progress";
  process: string;
  current: number;
  target: number;
}

export type Body = OneShotBody | ProgressBody;

/**
 * factory function to create one-shot notification
 * @param id
 * @param level
 * @param message
 */
export function createOneShot(id: string, level: Level, message: string) {
  return new Notification(id, level, { kind: NotificationType.OneShot, message });
}

/**
 * factory function to create progress notification
 * @param id
 * @param level
 * @param progress
 */
export function createProgress(
  id: string,
  level: Level,
  progress: { process: string; current: number; target: number }
) {
  return new Notification(id, level, { kind: NotificationType.Progress, ...progress });
}

// notification from server
export class Notification {
  constructor(public readonly id: string, public readonly level: Level, public readonly body: Body) {}
}
